xADDomain: Set-TargetResource Throws Terminating Error When DC Promotion in Progress

The first time DSC tries to install xADDomain, it runs Install-ADDSForrest (or Install-ADDSDomain but I haven't tried that one) appropriately.

If DSC comes back again to run this resource while the domain is still starting up (which happens in the Azure DSC extension), the Get method (which is run by the Test method) will catch the resulting ADServerDownException error and then return nothing.
This triggers the resource to call the Set method again which calls Install-ADDSForrest (or Install-ADDSDomain) again.

This second call to Install-ADDSForrest spits out an ugly error:
[ERROR] Verification of prerequisites for Domain Controller promotion failed. The specified argument 'DataBasePath' was not recognized

This causes the entire configuration to fail and nothing that depends on this resource can run.
But the call to Install-ADDSForrest has actually already happened, so the machine is eventually promoted to a domain controller successfully.

This issue is breaking a lot of ARM templates with AD in Windows Server 2016 since it seems to take longer for DC promotion to occur.

A quick work-around to this problem is to retry retrieving the domain in the Get method until it is found or the resource hits a timeout.

I'm getting some long delays (5+ minutes) promoting Windows Server 2016 TP4 to a DC using xADDomain. It does eventually get there though.

I have not seen this error though:
[ERROR] Verification of prerequisites for Domain Controller promotion failed. The specified argument 'DataBasePath' was not recognized.

We also encounter this issue extensively running 2012R2 with WMF & DSC Extension on Azure.

On the second reboot, it find the domain and I get:
VERBOSE: [2016-06-15T19:30:41] [VERBOSE] [SLAPPDEVDC0]: LCM:  [ Start  Resource ]  
[[cADDomain]FirstDS::[cPrimaryDomainController]primaryDc]
VERBOSE: [2016-06-15T19:30:41] [VERBOSE] [SLAPPDEVDC0]: LCM:  [ Start  Test     ]  
[[cADDomain]FirstDS::[cPrimaryDomainController]primaryDc]
VERBOSE: [2016-06-15T19:30:43] [VERBOSE] [SLAPPDEVDC0]:                            
[[cADDomain]FirstDS::[cPrimaryDomainController]primaryDc] Active Directory domain 'slappdev.cloud' found.
VERBOSE: [2016-06-15T19:30:43] [VERBOSE] [SLAPPDEVDC0]: LCM:  [ End    Test     ]  
[[cADDomain]FirstDS::[cPrimaryDomainController]primaryDc]  in 2.2660 seconds.
VERBOSE: [2016-06-15T19:30:43] [VERBOSE] [SLAPPDEVDC0]: LCM:  [ Skip   Set      ]  
[[cADDomain]FirstDS::[cPrimaryDomainController]primaryDc]
VERBOSE: [2016-06-15T19:30:43] [VERBOSE] [SLAPPDEVDC0]: LCM:  [ End    Resource ]  

on the third reboot of my sequence It usually don't find the domain:
[[cADDomain]FirstDS::[cPrimaryDomainController]primaryDc]
VERBOSE: [2016-06-15T19:33:00] [VERBOSE] [SLAPPDEVDC0]: LCM:  [ Start  Test     ]  
[[cADDomain]FirstDS::[cPrimaryDomainController]primaryDc]
VERBOSE: [2016-06-15T19:33:01] [VERBOSE] [SLAPPDEVDC0]:                            
[[cADDomain]FirstDS::[cPrimaryDomainController]primaryDc] Domain '' is NOT present on the current node.
VERBOSE: [2016-06-15T19:33:01] [VERBOSE] [SLAPPDEVDC0]: LCM:  [ End    Test     ]  
[[cADDomain]FirstDS::[cPrimaryDomainController]primaryDc]  in 1.0310 seconds.
VERBOSE: [2016-06-15T19:33:01] [VERBOSE] [SLAPPDEVDC0]: LCM:  [ Start  Set      ]  
[[cADDomain]FirstDS::[cPrimaryDomainController]primaryDc]
VERBOSE: [2016-06-15T19:33:01] [VERBOSE] [SLAPPDEVDC0]:                            
[[cADDomain]FirstDS::[cPrimaryDomainController]primaryDc] Creating AD forest 'slappdev.cloud' ...
VERBOSE: [2016-06-15T19:33:01] [ERROR] Verification of prerequisites for Domain Controller promotion failed. The specified argument 'DataBasePath' was not recognized.

VERBOSE: [2016-06-15T19:33:01] [VERBOSE] [SLAPPDEVDC0]:                            
[[cADDomain]FirstDS::[cPrimaryDomainController]primaryDc] AD forest '{0}' created.
VERBOSE: [2016-06-15T19:33:01] [VERBOSE] [SLAPPDEVDC0]: LCM:  [ End    Set      ]  
[[cADDomain]FirstDS::[cPrimaryDomainController]primaryDc]  in 0.6100 seconds.
VERBOSE: [2016-06-15T19:33:01] [ERROR] The PowerShell DSC resource '[cADDomain]FirstDS::[cPrimaryDomainController]primaryDc' 
with SourceInfo 'C:\Program 
Files\WindowsPowerShell\Modules\cOrckestra\DSCResources\cPrimaryDomainController\cPrimaryDomainController.schema.psm1::47::5::c
ADDomain' threw one or more non-terminating errors while running the Set-TargetResource functionality. These errors are logged 
to the ETW channel called Microsoft-Windows-DSC/Operational. Refer to this channel for more details.
VERBOSE: [2016-06-15T19:33:01] [VERBOSE] [SLAPPDEVDC0]: LCM:  [ End    Set      ]
VERBOSE: [2016-06-15T19:33:01] [ERROR] The SendConfigurationApply function did not succeed.
VERBOSE: [2016-06-15T19:33:01] [VERBOSE] Operation 'Invoke CimMethod' complete.
VERBOSE: [2016-06-15T19:33:01] [VERBOSE] Time taken for configuration job to complete is 21.792 seconds

I know you see cADDomain as the resource but it is xADDomain with one more parameter that we added for DNS.

--- a/DSCResources/MSFT_xADDomain/MSFT_xADDomain.psm1
+++ b/DSCResources/MSFT_xADDomain/MSFT_xADDomain.psm1
@@ -20,18 +20,47 @@ data localizedData
         ResourceInDesiredState               = Resource '{0}' is in the desired state.
         ResourceNotInDesiredState            = Resource '{0}' is NOT in the desired state.
         RetryingGetADDomain                  = Attempt {0} of {1} to call Get-ADDomain failed, retrying in {2} seconds.
-    UnhandledError                       = Unhandled error occured, detail here: {0}
-    FaultExceptionAndDomainShouldExist = ServiceModel FaultException detected and domain should exist, performing retry...
-
+        UnhandledError                       = Unhandled error occured, detail here: {0}
+        FaultExceptionAndDomainShouldExist = ServiceModel FaultException detected and domain should exist, performing retry...
 '@
 }
 
+<#
+    .SYNOPSIS
+        Retrieves the name of the file that tracks the status of the xADDomain resource with the
+        specified domain name.
+
+    .PARAMETER DomainName
+        The domain name of the xADDomain resource to retrieve the tracking file name of.
+
+    .NOTES
+        The tracking file is currently output to the environment's temp directory.
+        
+        This file is NOT removed when a configuration completes, so if another call to a xADDomain
+        resource with the same domain name occurs in the same environment, this file will already
+        be present.
+        
+        This is so that when another call is made to the same resource, the resource will not
+        attempt to promote the machine to a domain controller again (which would cause an error).
+        
+        If the resource should be promoted to a domain controller once again, you must first remove
+        this file from the environment's temp directory (usually C:\Temp).
+
+        If in the future this functionality needs to change so that future configurations are not
+        affected, $env:temp should be changed to the resource's cache location which is removed
+        after each configuration.
+        ($env:systemRoot\system32\Configuration\BuiltinProvCache\MSFT_xADDomain)
+#>
 function Get-TrackingFilename {
-param(
-    [Parameter(Mandatory)]
-    [String] $DomainName
+    [OutputType([String])]
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory = $true)]
+        [String]
+        $DomainName
     )
-    Join-Path $Env:TEMP ('{0}.xADDomain.completed' -f $DomainName)
+
+    return Join-Path -Path ($env:temp) -ChildPath ('{0}.xADDomain.completed' -f $DomainName)
 }
 
 function Get-TargetResource
@@ -135,7 +164,7 @@ function Get-TargetResource
     if($domainShouldExist) {
         $retries++
         Write-Verbose ($localizedData.RetryingGetADDomain -f $retries, $maxRetries, $retryIntervalInSeconds)
-        Sleep -Seconds ($retries * $retryIntervalInSeconds)
+        Start-Sleep -Seconds ($retries * $retryIntervalInSeconds)
     }
 
     } while ($domainShouldExist -and ($retries -le $maxRetries) )

--- a/DSCResources/MSFT_xADDomain/MSFT_xADDomain.psm1
+++ b/DSCResources/MSFT_xADDomain/MSFT_xADDomain.psm1
@@ -7,7 +7,7 @@ data localizedData
         InvalidDomainError                   = Computer is a member of the wrong domain?!
         ExistingDomainMemberError            = Computer is already a domain member. Cannot create a new '{0}' domain?
         InvalidCredentialError               = Domain '{0}' is available, but invalid credentials were supplied.
-        							         
+                                             
         QueryDomainWithLocalCredential       = Computer is a domain member; querying domain '{0}' using local credential ...
         QueryDomainWithCredential            = Computer is a workgroup member; querying for domain '{0}' using supplied credential ...
         DomainFound                          = Active Directory domain '{0}' found.
@@ -20,18 +20,18 @@ data localizedData
         ResourceInDesiredState               = Resource '{0}' is in the desired state.
         ResourceNotInDesiredState            = Resource '{0}' is NOT in the desired state.
         RetryingGetADDomain                  = Attempt {0} of {1} to call Get-ADDomain failed, retrying in {2} seconds.
-	UnhandledError                       = Unhandled error occured, detail here: {0}
-	FaultExceptionAndDomainShouldExist = ServiceModel FaultException detected and domain should exist, performing retry...
+    UnhandledError                       = Unhandled error occured, detail here: {0}
+    FaultExceptionAndDomainShouldExist = ServiceModel FaultException detected and domain should exist, performing retry...
 
 '@
 }
 
 function Get-TrackingFilename {
 param(
-	[Parameter(Mandatory)]
+    [Parameter(Mandatory)]
     [String] $DomainName
-	)
-	Join-Path $Env:TEMP ('{0}.xADDomain.completed' -f $DomainName)
+    )
+    Join-Path $Env:TEMP ('{0}.xADDomain.completed' -f $DomainName)
 }
 
 function Get-TargetResource
@@ -75,68 +75,68 @@ function Get-TargetResource
     $maxRetries = 5
     $retryIntervalInSeconds = 30
     $domainShouldExist = (Test-Path (Get-TrackingFilename -DomainName $DomainName))
-    do {			
-	try
-	{
-		if ($isDomainMember) {
-			## We're already a domain member, so take the credentials out of the equation
-			Write-Verbose ($localizedData.QueryDomainADWithLocalCredentials -f $domainFQDN);
-			$domain = Get-ADDomain -Identity $domainFQDN -ErrorAction Stop;
-		}
-		else {
-			Write-Verbose ($localizedData.QueryDomainWithCredential -f $domainFQDN);
-			$domain = Get-ADDomain -Identity $domainFQDN -Credential $DomainAdministratorCredential -ErrorAction Stop;
-		}
+    do {            
+    try
+    {
+        if ($isDomainMember) {
+            ## We're already a domain member, so take the credentials out of the equation
+            Write-Verbose ($localizedData.QueryDomainADWithLocalCredentials -f $domainFQDN);
+            $domain = Get-ADDomain -Identity $domainFQDN -ErrorAction Stop;
+        }
+        else {
+            Write-Verbose ($localizedData.QueryDomainWithCredential -f $domainFQDN);
+            $domain = Get-ADDomain -Identity $domainFQDN -Credential $DomainAdministratorCredential -ErrorAction Stop;
+        }
 
-		## No need to check whether the node is actually a domain controller. If we don't throw an exception,
-		## the domain is already UP - and this resource shouldn't run. Domain controller functionality
-		## should be checked by the xADDomainController resource?
-		Write-Verbose ($localizedData.DomainFound -f $domain.DnsRoot);
+        ## No need to check whether the node is actually a domain controller. If we don't throw an exception,
+        ## the domain is already UP - and this resource shouldn't run. Domain controller functionality
+        ## should be checked by the xADDomainController resource?
+        Write-Verbose ($localizedData.DomainFound -f $domain.DnsRoot);
         
-		$targetResource = @{
-			DomainName = $domain.DnsRoot;
-			ParentDomainName = $domain.ParentDomain;
-			DomainNetBIOSName = $domain.NetBIOSName;
-		}
+        $targetResource = @{
+            DomainName = $domain.DnsRoot;
+            ParentDomainName = $domain.ParentDomain;
+            DomainNetBIOSName = $domain.NetBIOSName;
+        }
         
-		return $targetResource;
-	}
-	catch [Microsoft.ActiveDirectory.Management.ADIdentityNotFoundException]
-	{
-		$errorMessage = $localizedData.ExistingDomainMemberError -f $DomainName;
-		ThrowInvalidOperationError -ErrorId 'xADDomain_DomainMember' -ErrorMessage $errorMessage;
-	}
-	catch [Microsoft.ActiveDirectory.Management.ADServerDownException]
-	{
-		Write-Verbose ($localizedData.DomainNotFound -f $domainFQDN)
-		$domain = @{ };
-		# will fall into retry mechanism
-	}
-	catch [System.Security.Authentication.AuthenticationException]
-	{
-		$errorMessage = $localizedData.InvalidCredentialError -f $DomainName;
-		ThrowInvalidOperationError -ErrorId 'xADDomain_InvalidCredential' -ErrorMessage $errorMessage;
-	}
-	catch
-	{
-		$errorMessage = $localizedData.UnhandledError -f ($_.Exception | Format-List -Force | Out-String)
-		Write-Verbose $errorMessage
+        return $targetResource;
+    }
+    catch [Microsoft.ActiveDirectory.Management.ADIdentityNotFoundException]
+    {
+        $errorMessage = $localizedData.ExistingDomainMemberError -f $DomainName;
+        ThrowInvalidOperationError -ErrorId 'xADDomain_DomainMember' -ErrorMessage $errorMessage;
+    }
+    catch [Microsoft.ActiveDirectory.Management.ADServerDownException]
+    {
+        Write-Verbose ($localizedData.DomainNotFound -f $domainFQDN)
+        $domain = @{ };
+        # will fall into retry mechanism
+    }
+    catch [System.Security.Authentication.AuthenticationException]
+    {
+        $errorMessage = $localizedData.InvalidCredentialError -f $DomainName;
+        ThrowInvalidOperationError -ErrorId 'xADDomain_InvalidCredential' -ErrorMessage $errorMessage;
+    }
+    catch
+    {
+        $errorMessage = $localizedData.UnhandledError -f ($_.Exception | Format-List -Force | Out-String)
+        Write-Verbose $errorMessage
 
-		if ($domainShouldExist -and ($_.Exception.InnerException -is [System.ServiceModel.FaultException]))
-		{
-			Write-Verbose $localizedData.FaultExceptionAndDomainShouldExist
-			# will fall into retry mechanism
-		} else {
-			## Not sure what's gone on here!
-			throw $_
-		}
-	}
+        if ($domainShouldExist -and ($_.Exception.InnerException -is [System.ServiceModel.FaultException]))
+        {
+            Write-Verbose $localizedData.FaultExceptionAndDomainShouldExist
+            # will fall into retry mechanism
+        } else {
+            ## Not sure what's gone on here!
+            throw $_
+        }
+    }
 
-	if($domainShouldExist) {
-		$retries++
-		Write-Verbose ($localizedData.RetryingGetADDomain -f $retries, $maxRetries, $retryIntervalInSeconds)
-		Sleep -Seconds ($retries * $retryIntervalInSeconds)
-	}
+    if($domainShouldExist) {
+        $retries++
+        Write-Verbose ($localizedData.RetryingGetADDomain -f $retries, $maxRetries, $retryIntervalInSeconds)
+        Sleep -Seconds ($retries * $retryIntervalInSeconds)
+    }
 
     } while ($domainShouldExist -and ($retries -le $maxRetries) )
 
@@ -303,7 +303,7 @@ function Set-TargetResource
         Write-Verbose -Message ($localizedData.CreatedForest -f $DomainName); 
     }  
 
-	"Finished" | Out-File -FilePath (Get-TrackingFilename -DomainName $DomainName) -Force
+    "Finished" | Out-File -FilePath (Get-TrackingFilename -DomainName $DomainName) -Force
 
     # Signal to the LCM to reboot the node to compensate for the one we
     # suppressed from Install-ADDSForest/Install-ADDSDomain

--- a/DSCResources/MSFT_xADDomain/MSFT_xADDomain.psm1
+++ b/DSCResources/MSFT_xADDomain/MSFT_xADDomain.psm1
@@ -3,23 +3,35 @@ data localizedData
 {
     # culture="en-US"
     ConvertFrom-StringData @'
-        RoleNotFoundError              = Please ensure that the PowerShell module for role '{0}' is installed.
-        InvalidDomainError             = Computer is a member of the wrong domain?!
-        ExistingDomainMemberError      = Computer is already a domain member. Cannot create a new '{0}' domain?
-        InvalidCredentialError         = Domain '{0}' is available, but invalid credentials were supplied.
-        
-        QueryDomainWithLocalCredential = Computer is a domain member; querying domain '{0}' using local credential ...
-        QueryDomainWithCredential      = Computer is a workgroup member; querying for domain '{0}' using supplied credential ...
-        DomainFound                    = Active Directory domain '{0}' found.
-        DomainNotFound                 = Active Directory domain '{0}' cannot be found.
-        CreatingChildDomain            = Creating domain '{0}' as a child of domain '{1}' ...
-        CreatedChildDomain             = Child domain '{0}' created.
-        CreatingForest                 = Creating AD forest '{0}' ...
-        CreatedForest                  = AD forest '{0}' created.
-        ResourcePropertyValueIncorrect = Property '{0}' value is incorrect; expected '{1}', actual '{2}'.
-        ResourceInDesiredState         = Resource '{0}' is in the desired state.
-        ResourceNotInDesiredState      = Resource '{0}' is NOT in the desired state.
+        RoleNotFoundError                    = Please ensure that the PowerShell module for role '{0}' is installed.
+        InvalidDomainError                   = Computer is a member of the wrong domain?!
+        ExistingDomainMemberError            = Computer is already a domain member. Cannot create a new '{0}' domain?
+        InvalidCredentialError               = Domain '{0}' is available, but invalid credentials were supplied.
+        							         
+        QueryDomainWithLocalCredential       = Computer is a domain member; querying domain '{0}' using local credential ...
+        QueryDomainWithCredential            = Computer is a workgroup member; querying for domain '{0}' using supplied credential ...
+        DomainFound                          = Active Directory domain '{0}' found.
+        DomainNotFound                       = Active Directory domain '{0}' cannot be found.
+        CreatingChildDomain                  = Creating domain '{0}' as a child of domain '{1}' ...
+        CreatedChildDomain                   = Child domain '{0}' created.
+        CreatingForest                       = Creating AD forest '{0}' ...
+        CreatedForest                        = AD forest '{0}' created.
+        ResourcePropertyValueIncorrect       = Property '{0}' value is incorrect; expected '{1}', actual '{2}'.
+        ResourceInDesiredState               = Resource '{0}' is in the desired state.
+        ResourceNotInDesiredState            = Resource '{0}' is NOT in the desired state.
+        RetryingGetADDomain                  = Attempt {0} of {1} to call Get-ADDomain failed, retrying in {2} seconds.
+	UnhandledError                       = Unhandled error occured, detail here: {0}
+	FaultExceptionAndDomainShouldExist = ServiceModel FaultException detected and domain should exist, performing retry...
+
 '@
+}
+
+function Get-TrackingFilename {
+param(
+	[Parameter(Mandatory)]
+    [String] $DomainName
+	)
+	Join-Path $Env:TEMP ('{0}.xADDomain.completed' -f $DomainName)
 }
 
 function Get-TargetResource
@@ -59,51 +71,74 @@ function Get-TargetResource
     $domainFQDN = Resolve-DomainFQDN -DomainName $DomainName -ParentDomainName $ParentDomainName;
     $isDomainMember = Test-DomainMember;
 
-    try
-    {
-        if ($isDomainMember) {
-            ## We're already a domain member, so take the credentials out of the equation
-            Write-Verbose ($localizedData.QueryDomainADWithLocalCredentials -f $domainFQDN);
-            $domain = Get-ADDomain -Identity $domainFQDN -ErrorAction Stop;
-        }
-        else {
-            Write-Verbose ($localizedData.QueryDomainWithCredential -f $domainFQDN);
-            $domain = Get-ADDomain -Identity $domainFQDN -Credential $DomainAdministratorCredential -ErrorAction Stop;
-        }
+    $retries = 0
+    $maxRetries = 5
+    $retryIntervalInSeconds = 30
+    $domainShouldExist = (Test-Path (Get-TrackingFilename -DomainName $DomainName))
+    do {			
+	try
+	{
+		if ($isDomainMember) {
+			## We're already a domain member, so take the credentials out of the equation
+			Write-Verbose ($localizedData.QueryDomainADWithLocalCredentials -f $domainFQDN);
+			$domain = Get-ADDomain -Identity $domainFQDN -ErrorAction Stop;
+		}
+		else {
+			Write-Verbose ($localizedData.QueryDomainWithCredential -f $domainFQDN);
+			$domain = Get-ADDomain -Identity $domainFQDN -Credential $DomainAdministratorCredential -ErrorAction Stop;
+		}
 
-        ## No need to check whether the node is actually a domain controller. If we don't throw an exception,
-        ## the domain is already UP - and this resource shouldn't run. Domain controller functionality
-        ## should be checked by the xADDomainController resource?
-        Write-Verbose ($localizedData.DomainFound -f $domain.DnsRoot);
+		## No need to check whether the node is actually a domain controller. If we don't throw an exception,
+		## the domain is already UP - and this resource shouldn't run. Domain controller functionality
+		## should be checked by the xADDomainController resource?
+		Write-Verbose ($localizedData.DomainFound -f $domain.DnsRoot);
         
-        $targetResource = @{
-            DomainName = $domain.DnsRoot;
-            ParentDomainName = $domain.ParentDomain;
-            DomainNetBIOSName = $domain.NetBIOSName;
-        }
+		$targetResource = @{
+			DomainName = $domain.DnsRoot;
+			ParentDomainName = $domain.ParentDomain;
+			DomainNetBIOSName = $domain.NetBIOSName;
+		}
         
-        return $targetResource;
-    }
-    catch [Microsoft.ActiveDirectory.Management.ADIdentityNotFoundException]
-    {
-        $errorMessage = $localizedData.ExistingDomainMemberError -f $DomainName;
-        ThrowInvalidOperationError -ErrorId 'xADDomain_DomainMember' -ErrorMessage $errorMessage;
-    }
-    catch [Microsoft.ActiveDirectory.Management.ADServerDownException]
-    {
-        Write-Verbose ($localizedData.DomainNotFound -f $domainFQDN)
-        $domain = @{ };
-    }
-    catch [System.Security.Authentication.AuthenticationException]
-    {
-        $errorMessage = $localizedData.InvalidCredentialError -f $DomainName;
-        ThrowInvalidOperationError -ErrorId 'xADDomain_InvalidCredential' -ErrorMessage $errorMessage;
-    }
-    catch
-    {
-        ## Not sure what's gone on here!
-        throw $_
-    }
+		return $targetResource;
+	}
+	catch [Microsoft.ActiveDirectory.Management.ADIdentityNotFoundException]
+	{
+		$errorMessage = $localizedData.ExistingDomainMemberError -f $DomainName;
+		ThrowInvalidOperationError -ErrorId 'xADDomain_DomainMember' -ErrorMessage $errorMessage;
+	}
+	catch [Microsoft.ActiveDirectory.Management.ADServerDownException]
+	{
+		Write-Verbose ($localizedData.DomainNotFound -f $domainFQDN)
+		$domain = @{ };
+		# will fall into retry mechanism
+	}
+	catch [System.Security.Authentication.AuthenticationException]
+	{
+		$errorMessage = $localizedData.InvalidCredentialError -f $DomainName;
+		ThrowInvalidOperationError -ErrorId 'xADDomain_InvalidCredential' -ErrorMessage $errorMessage;
+	}
+	catch
+	{
+		$errorMessage = $localizedData.UnhandledError -f ($_.Exception | Format-List -Force | Out-String)
+		Write-Verbose $errorMessage
+
+		if ($domainShouldExist -and ($_.Exception.InnerException -is [System.ServiceModel.FaultException]))
+		{
+			Write-Verbose $localizedData.FaultExceptionAndDomainShouldExist
+			# will fall into retry mechanism
+		} else {
+			## Not sure what's gone on here!
+			throw $_
+		}
+	}
+
+	if($domainShouldExist) {
+		$retries++
+		Write-Verbose ($localizedData.RetryingGetADDomain -f $retries, $maxRetries, $retryIntervalInSeconds)
+		Sleep -Seconds ($retries * $retryIntervalInSeconds)
+	}
+
+    } while ($domainShouldExist -and ($retries -le $maxRetries) )
 
 } #end function Get-TargetResource
 
@@ -265,8 +300,10 @@ function Set-TargetResource
             $installADDSParams['DomainNetbiosName'] = $DomainNetBIOSName;
         }
         Install-ADDSForest @installADDSParams;
-        Write-Verbose -Message ($localizedData.CreatedForest); 
+        Write-Verbose -Message ($localizedData.CreatedForest -f $DomainName); 
     }  
+
+	"Finished" | Out-File -FilePath (Get-TrackingFilename -DomainName $DomainName) -Force
 
     # Signal to the LCM to reboot the node to compensate for the one we
     # suppressed from Install-ADDSForest/Install-ADDSDomain

--- a/README.md
+++ b/README.md
@@ -245,6 +245,8 @@ Setting an ODJ Request file path for a configuration that creates a computer acc
 ### Unreleased
 
 * xAdDomainController: Update to complete fix for SiteName being required field.
+* xADDomain: Added retry logic to prevent FaultException to crash in Get-TargetResource on subsequent reboots after a domain is created because the service is not yet running. This error is mostly occur when the resource is used with the DSCExtension on Azure. 
+
 ### 2.15.0.0
 * xAdDomainController: Fixes SiteName being required field.
 

